### PR TITLE
Add more functionality for Python scripts…

### PR DESCRIPTION
**Pull request by davidcarroll**
Add more functionality for Python scripts (giving dashboards) to improve performance and cleaner Code

### CODE REVIEW

**Comment by davidcarroll** [comment]
This should be the last PR enabling me to publish to production the new Giving Dashboard recipe for churches.

**Comment by davidcarroll** [comment]
That last commit was to be able to search ContributionTags too.
This makes it easier to validate that all contributions are accounted for and that no contributions slipped through the cracks.

File: CmsData/Migrations/20181022_update_Contributions2_func.sql [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	join dbo.BundleHeader h on d.BundleHeaderId = h.BundleHeaderId
+	join lookup.BundleHeaderTypes bht on h.BundleHeaderTypeId = bht.Id
+	join lookup.BundleStatusTypes bst on h.BundleStatusId = bst.Id
+	join dbo.People p on c.PeopleId = p.PeopleId
+	join dbo.Families fa on p.FamilyId = fa.FamilyId
+	join dbo.People sp on sp.PeopleId = p.SpouseId
+	join lookup.MaritalStatus mssp on mssp.Id = sp.MaritalStatusId
+	join lookup.MaritalStatus msp on msp.Id = p.MaritalStatusId
+	left outer join dbo.[Transaction] t on t.Id = c.TranId
+where 1 = 1
```
[the remark this answers is not in the document]
Comment (davidcarroll) [reply CmsData/Migrations/20181022_update_Contributions2_func.sql]: I do that sometimes when I am debugging and need to comment out the clauses one by one to see the effects. Makes it easier to not have to comment out the first clause and only have to deal with the clauses that begin with AND. Feel free to remove any that you see left in if they bother you. I’ll try to remember to fix any of my future code commits.

File: CmsData/Migrations/20181013_update_contributions2_func.sql [hunk 2]
```diff
@@ -21,13 +21,13 @@ SELECT
...
 			AND ISNULL(sp.ContributionOptionsId, CASE WHEN mssp.Married = 1 THEN 2 ELSE 1 END) = 2
-			AND ISNULL(p.ContributionOptionsId, CASE WHEN ms.Married = 1 THEN 2 ELSE 1 END) = 2
+			AND ISNULL(p.ContributionOptionsId, CASE WHEN msp.Married = 1 THEN 2 ELSE 1 END) = 2
 		THEN sp.PeopleId 
 		ELSE c.PeopleId 
 	END AS CreditGiverId,
 
     CASE WHEN ISNULL(sp.ContributionOptionsId, CASE WHEN mssp.Married = 1 THEN 2 ELSE 1 END) = 1
-			OR ISNULL(p.ContributionOptionsId, CASE WHEN ms.Married = 1 THEN 2 ELSE 1 END) = 1
+			OR ISNULL(p.ContributionOptionsId, CASE WHEN msp.Married = 1 THEN 2 ELSE 1 END) = 1
```
Comment (davidcarroll) [inline CmsData/Migrations/20181013_update_contributions2_func.sql]: @aternex we were both fixing the typo but you changed this part and I had changed the join statement. Thus it was still broken since we were fixing two different parts. Drove me crazy trying to figure out what was going on!

**Comment by davidcarroll** [comment]
@bron, @aternex I am waiting on getting this to production since it is required to install the Giving Dashboard into production for the client. Let me know if I need to do anything to facilitate.

**Comment by davidcarroll** [comment]
I never thought to test the Migration I was adding by running a database build.
There were problems during testing because of having multiple scripts on the same day so the sort order was not in dependency order. I renamed the scripts so that they would be run in the proper order.

BTW, this never was a problem with RedGate SQL compare since it always took care of the dependency order of scripts and it was all machine generated code. Having to manually create and test the migration scripts is more work and more error prone.

**Comment by aternex** [comment]
It might be irrelevant, but we have added the functionality to limit access to funds based on role membership. It looks as if the sql here would pull all contributions. I feel as if those using this functionality would (or should) have access to any data anyway. In the spots we are using this we have done it in app vs in sql directly (more or less). However, really it just looks at the current user and gets a list of their roles and deals with those FinanceViewOnly users in a special case. What would the desired functionality be if a user looking at the GivingDashboard was limited to see only specific funds. Would we simply limit their ability to see this Dashboard or would we need to add similar logic to these methods that pull contribution data?

In other words, if a user is FinanceViewOnly and the church has setup a fund that has a role manager id then anywhere inside the app that user goes to view contribution details or statement details the source data is scoped to only those roles he or she is granted access to. For Finance users, they just see everything anyway.

**Comment by davidcarroll** [comment]
Thank you for your comment and working to make this a priority. I am not used to waiting weeks on getting commits pulled into master and into production. I do understand why. It does make it a very slow process to get work done for a waiting client when I said it would just take a few days. I am so used to being able to publish to production on an as needed basis. Easy to do when you are the only developer! 

The Python functions related to contributions only create tags for performance and caching for the Giving Dashboard. The Giving Dashboard, written in SQL and Python will only display summary totals. It is the custom SQL code that gives the power to retrieve data, read only, from the entire database. That functionality has been there since we created the ability to create custom reports written in SQL. It is up to the client to choose how to protect those reports with a meta comment that allows only certain roles to view.

Adding support for the limited funds view for a special role is out of the scope of the Giving Dashboard. Because no detailed individual contributions will be seen, the summary totals should not be a problem for the FinanceViewOnly role.
